### DISCUSSION TITLE
Preserving '_StereochemDone' and other mol/atom props after pickle/dill from multithreading?

### DISCUSSION THREAD

**Discussion by IsaacJK** [rdkit/rdkit#4137] [Q&A] [answered]
Python 3.8.8, rdkit.__version__ = 2021.03.2, anaconda environment on Ubuntu 18

From below, seems like passing back mol from thread via pickle (multiprocessing) or dill (multiprocess/pathos) strips mol/atom properties.  Haven't looked beyond '_StereochemDone' and '_ChiralityPossible', as it was the first failure point in using rdMolDescriptors.CalcNumAtomStereoCenters.

Does this seem like an expected unavoidable consequence of boost vs python multithreading issues, or fixable by spme way I'm missing other than manually passing back the properties via dict and resetting them afterwards?

Thanks!

```
from rdkit.Chem import MolFromSmiles
# import multiprocessing as mp
# import multiprocess as mp
from pathos import multiprocessing as mp
from functools import partial

# single
smiles = "CC[C@H](C)Br"
m1 = MolFromSmiles(smiles)
print("Single thread, no pkl")
try:
    print("  _StereochemDone=", m1.GetProp("_StereochemDone"))
except KeyError:
    print("  Invalid Key")
print(" "*3, [a.HasProp('_ChiralityPossible') for a in m1.GetAtoms()])

# multi
nthread = mp.cpu_count()
def mol_from_smiles(smiles):
    mol = MolFromSmiles(smiles)
    print("In thread")
    try:
        print("  _StereochemDone=", mol.GetProp("_StereochemDone"))
    except KeyError:
        print("  Invalid Key")
    print(" "*3, [a.HasProp('_ChiralityPossible') for a in mol.GetAtoms()])
    return mol

# Generate fingerprint vectors in parallel threading
with mp.Pool(nthread) as pool:
    results = pool.map(partial(mol_from_smiles, ), [smiles])
m2 = results[0]
print("After mp")
try:
    print("  _StereochemDone=", m2.GetProp("_StereochemDone"))
except KeyError:
    print("Invalid Key")
print(" "*3, [a.HasProp('_ChiralityPossible') for a in m2.GetAtoms()])

Output:
Single thread, no pkl
  _StereochemDone= 1 
   [0, 0, 1, 0, 0]
In thread
  _StereochemDone= 1
   [0, 0, 1, 0, 0]
After mp
Invalid Key
   [0, 0, 0, 0, 0]

**Comment by ptosco** [comment, marked as answer]
@IsaacJK Try setting
```
Chem.SetDefaultPickleProperties(Chem.PropertyPickleOptions.MolProps)
```
If that still fails, try returning an explicitly pickled molecule rather then delegating the pickling to `multiprocessing`.

**Reply by IsaacJK** [reply]
Thanks!

MolProps didn't work, but AllProps did
`Chem.SetDefaultPickleProperties(Chem.PropertyPickleOptions.AllProps)`

Interestingly (maybe?) all permutations of: mol, PropertyMol(mol), and/or passing explicit pickle object (pickle, cPickle ... multiprocess, multiprocessing, etc) didn't work.
